providers/ldap/ldap_provider.go

```diff
@@ -0,0 +1,650 @@
+	ldapAccessToken                 = name + "AccessToken"
+	ldapJwt                         = name + "Jwt"
+	memberOfAttribute               = "memberOf"
+	objectClassAttribute            = "objectClass"
+	settingBase                     = "api.auth.ldap."
+	accessModeSetting               = settingBase + "access.mode"
+	domainSetting                   = settingBase + "domain"
+	groupSearchDomainSetting        = settingBase + "group.search.domain"
+	loginDomainSetting              = settingBase + "login.domain"
+	portSetting                     = settingBase + "port"
+	userSearchFieldSetting          = settingBase + "user.search.field"
+	serviceAccountUsernameSetting   = settingBase + "service.account.user"
+	groupSearchFieldSetting         = settingBase + "group.search.field"
+	userObjectClassSetting          = settingBase + "user.object.class"
+	userNameFieldSetting            = settingBase + "user.name.field"
+	groupObjectClassSetting         = settingBase + "group.object.class"
+	userLoginFieldSetting           = settingBase + "user.login.field"
+	userDisabledBitMaskSetting      = settingBase + "user.enabled.mask.bit"
+	serverSetting                   = settingBase + "server"
+	serviceAccountPasswordSetting   = settingBase + "service.account.password"
+	userEnabledAttributeSetting     = settingBase + "user.enabled.attribute"
+	groupNameFieldSetting           = settingBase + "group.name.field"
+	tlsSetting                      = settingBase + "tls"
+	timeoutSetting                  = settingBase + "connection.timeout"
+	allowedIdentitiesSetting        = settingBase + "allowed.identities"
+	groupDnFieldSetting             = settingBase + "group.dn.field"
+	groupMemberUserAttributeSetting = settingBase + "group.member.user.attribute"
+)
+
+var apiClient *client.RancherClient
+var scopes = []string{userScope, groupScope}
+var nilIdentity = client.Identity{Resource: client.Resource{
+	Type: "identity",
+}}
+var entryAttributes []*ldap.EntryAttribute
+
+//InitializeProvider returns a new instance of the provider
+func InitializeProvider() *LdapProvider {
+	client := &http.Client{}
+	ldapClient := &LClient{}
+	ldapClient.httpClient = client
+
+	ldapProvider := &LdapProvider{}
+	ldapProvider.ldapClient = ldapClient
+
+	return ldapProvider
+}
+
+type LdapProvider struct {
+	ldapClient *LClient
+}
+
+func (l *LdapProvider) GetName() string {
+	return name
+}
+
+func (l *LdapProvider) GetUserType() string {
+	return userScope
+}
+
+func (l *LdapProvider) GenerateToken(jsonInput map[string]string) (model.Token, error) {
+	log.Info("Now generating Ldap token")
+	split := strings.Split(jsonInput["code"], ":")
+	username, password := split[0], split[1]
+	externalId := username
+	if l.ldapClient.config.LoginDomain != "" {
+		externalId = l.ldapClient.config.LoginDomain + "\\" + username
+	}
+
+	if password == "" {
+		return model.Token{}, fmt.Errorf("Failed to login, password not provided")
+	}
+
+	ldapServer := l.ldapClient.config.Server
```

Comment:
For better readability and error handling/logic separation:
Add a new class to do all the work w.r.t l.ldapClient. This class should handle all communication with backend ldap, read the responses and return the ldap user/group account data
Keep ldap_provider as an interface between auth_server and the above class. ldap_provider should translate from ldap user data to  rancher identity.
Checkout github implementation.